File: data/code_pairs/fortran/2403.f90
! translation.f90
module types_mod
    implicit none

    type :: t0
        integer :: j = 42
    end type t0

    type :: t
        integer :: i
        class(t0), allocatable :: foo
    end type t
end module types_mod

program translation
    use types_mod
    implicit none

    type(t) :: m

    allocate(t0 :: m%foo)
    m%i = 5

    select type(bar => m%foo)
    type is (t0)
        print *, "Fortran test: bar%j =", bar%j
    end select
end program translation

File: data/code_pairs/cpp/2403.cpp
// translation.cpp
#include <iostream>
#include <memory>

class T0 {
public:
    int j;
    T0() : j(42) {}
};

class T {
public:
    int i;
    std::unique_ptr<T0> foo;

    T() : i(0), foo(nullptr) {}
};

int main() {
    T m;
    m.foo = std::make_unique<T0>();
    m.i = 5;

    if (m.foo) {
        T0 &bar = *m.foo;
        std::cout << "C++ test: bar.j = " << bar.j << std::endl;
    }

    return 0;
}